server/py/framework/utils/clients/discovery.py

```diff
@@ -0,0 +1,150 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import collections
+import re
+import typing
+from dataclasses import dataclass
+
+import mlrun.utils.singleton
+from mlrun import mlconf
+
+
+@dataclass
+class ServiceInstance:
+    name: str
+    url: str
+    method_routes: dict[str, list[re.Pattern]] = None
+    status: str = "UP"  # UP, DOWN, UNKNOWN
+
+
+class Client(
+    metaclass=mlrun.utils.singleton.AbstractSingleton,
+):
+    def __init__(self):
+        self._logger = mlrun.utils.logger.get_child(__name__)
+        self.services = None
+        self.initialize()
+
+    def initialize(self):
+        # We use an ordered dict for control over service matching order.
+        # This is important for services that may have overlapping routes.
+        self.services: dict[str, ServiceInstance] = collections.OrderedDict()
+        if mlconf.services.hydra.services != "*":
+            self.register_service(service_name="alerts")
+        self.register_service(service_name="api")
+        # Chief is last to allow other services to override its routes
+        self.register_service(service_name="api-chief")
+
+    def register_service(
+        self,
+        service_name: str,
+    ):
+        """Register a new service instance."""
+        method_routes = self._resolve_service_method_routes(service_name)
+        url = self._resolve_service_url(service_name)
+        self.services[service_name] = ServiceInstance(
+            name=service_name,
+            url=url,
+            method_routes=method_routes,
+        )
+        self._logger.info("Registered service", service_name=service_name, url=url)
+
+    def deregister_service(self, service_name: str):
+        """Deregister a service instance."""
+        if service_name in self.services:
+            del self.services[service_name]
+
+        self._logger.info(
+            "Deregistered service",
+            service_name=service_name,
+        )
+
+    def resolve_service_by_request(
+        self, method: str, path: str
+    ) -> typing.Optional[ServiceInstance]:
+        """
+        Resolve path and returns the matching service instance for the request.
+
+        :param method: HTTP method of the request
+        :param path: URL path to match against service patterns
+        :return: ServiceInstance matching the request or None if no match
+        """
+        method = method.lower()
+        if service_name := self._find_service(method, path):
+            return self.get_service(service_name)
+        return None
+
+    def get_service(self, service_name: str) -> typing.Optional[ServiceInstance]:
+        """Get all healthy instances of a service."""
```

Comment:
Isn't the services dict values are only one service per key?
If so, what is the meaning of all healthy instances of a service?